torchrl/objectives/dqn.py

```diff
@@ -198,9 +206,13 @@ def forward(self, input_tensordict: TensorDictBase) -> TensorDict:
             buffers=self.value_network_buffers,
         )  # Log probabilities log p(s_t, ·; θonline)
         action_log_softmax = td_clone.get("action_value")
-        action_expand = action.unsqueeze(-2).expand_as(action_log_softmax)
-        log_ps_a = action_log_softmax.masked_select(action_expand.to(torch.bool))
-        log_ps_a = log_ps_a.view(batch_size, atoms)  # log p(s_t, a_t; θonline)
+
+        if _CATEGORICAL_ACTION_ENCODING:
+            log_ps_a = action_log_softmax[range(batch_size), :, action.squeeze(-1)]
```

Comment:
same comments as above regarding indexing, usage of env variable and test coverage

Reply:
Reworked this as well with a few tweaks since we work with atoms here.